fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/ProtocolLogicalMetricGenerator.java

```diff
@@ -0,0 +1,177 @@
+    }
+
+    /**
+     * Extracts the list of metrics from the url metric query string and generates a set of LogicalMetrics.
+     * <p>
+     * If the query contains undefined metrics, {@link BadApiRequestException} will be
+     * thrown.
+     *
+     * This method is meant for backwards compatibility. If you do not need to use this method for that reason please
+     * prefer using a generator instance instead.
+     *
+     * @param apiMetricQuery  URL query string containing the metrics separated by ','
+     * @param metricDictionary  Metric dictionary contains the map of valid metric names and logical metric objects
+     *
+     * @return set of metric objects
+     */
+    public LinkedHashSet<LogicalMetric> generateLogicalMetrics(
+            String apiMetricQuery,
+            MetricDictionary metricDictionary
+    ) {
+        LinkedHashSet<LogicalMetric> metrics = new LinkedHashSet<>();
+        List<String> invalidMetricNames = new ArrayList<>();
+
+        List<ApiMetric> apiMetrics = protocolAntlrApiMetricParser.apply(apiMetricQuery);
+
+
+        for (ApiMetric metric : apiMetrics) {
+            LogicalMetric logicalMetric = metricDictionary.get(metric.getBaseApiMetricId());
+//            protocolChain.applyProtocols();
+            if (logicalMetric == null) {
+                invalidMetricNames.add(metric.getRawName());
+                continue;
+            }
+            logicalMetric = protocolChain.applyProtocols(
+                    new LogicalMetricInfo(metric.getRawName()),
+                    metric,
+                    logicalMetric
+            );
+            metrics.add(logicalMetric);
+        }
+        if (!invalidMetricNames.isEmpty()) {
+            String message = ErrorMessageFormat.METRICS_UNDEFINED.logFormat(invalidMetricNames);
+            LOG.error(message);
+            throw new BadApiRequestException(message);
+        }
+        return metrics;
+    }
+
+    /**
+     * Validate that all metrics are part of the logical table.
+     *
+     * This method is meant for backwards compatibility. If you do not need to use this method for that reason please
+     * prefer using a generator instance instead.
+     *
+     * @param logicalMetrics  The set of metrics being validated
+     * @param table  The logical table for the request
+     *
+     * @throws BadApiRequestException if the requested metrics are not in the logical table
+     */
+    public static void validateMetrics(Set<LogicalMetric> logicalMetrics, LogicalTable table)
+            throws BadApiRequestException {
+        //get metric names from the logical table
+        Set<String> validMetricNames = table.getLogicalMetrics().stream()
+                .map(LogicalMetric::getName)
+                .collect(Collectors.toSet());
+
+        //get metric names from logicalMetrics and remove all the valid metrics
+        Set<String> invalidMetricNames = logicalMetrics.stream()
```

Comment:
This validation will break protocol metrics. Line 131 you are using the raw name as the output name for the logical metrics. Unless I am missing some other processing stage, the raw name will NOT be on the logical table.

Reply:
Fixed above